File: array/0532_k_diff_pairs_in_an_array/0532_k_diff_pairs_in_an_array.py
# -*- coding: utf-8 -*-


class Solution(object):
    def findPairs(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        if k < 0: return 0
        result, lookup = set(), set()
        for num in nums:
            if num-k in lookup:
                result.add(num-k)
            if num+k in lookup:
                result.add(num)
            lookup.add(num)
        return len(result)


nums = [3,1,4,1,5]
k = 2

print(Solution().findPairs(nums, k))

File: array/0532_k_diff_pairs_in_an_array/0532_k_diff_pairs_in_an_array.go
package main

import (
	"fmt"
)

func findPairs(nums []int, k int) int {
	if k < 0 {
		return 0
	}

	record := make(map[int]int)
	for _, num := range nums {
		record[num]++
	}

	ans := 0

	if k == 0 {
		for _, count := range record {
			if count > 1 {
				ans++
			}
		}
		return ans
	}

	for n := range record {
		if record[n-k] > 0 {
			ans++
		}
	}

	return ans
}
func main() {
	nums := []int{3, 1, 4, 1, 5}
	k := 2
	fmt.Println(findPairs(nums, k))
}
